src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -188,35 +189,105 @@ public static boolean isStringSafe(@CheckForNull String input, @CheckForNull Env
     ));
 
     /**
-     * Does first-level sanitization of a step's params, returning the Describable arguments map,
+     * Sanitize a list recursively
+     */
+    @CheckForNull
+    Object sanitizeListAndRecordMutation(@Nonnull List objects, @CheckForNull EnvVars variables) {
+        // Package scoped so we can test it directly
+
+        if (isOversized(objects)) {
+            this.isUnmodifiedBySanitization = false;
+            return NotStoredReason.OVERSIZE_VALUE;
+        }
+
+        boolean isMutated = false;
+        List output = new ArrayList(objects.size());
+        for (Object o : objects) {
+            Object modded = sanitizeObjectAndRecordMutation(o, variables);
+
+            if (modded != o) {
+                // Sanitization stripped out some values, so we need to store the mutated object
+                output.add(modded);
+                isMutated = true; //isUnmodifiedBySanitization was already set
+            } else { // Any mutation was just from exploding step/uninstantiated describable, and we can just use the original
+                output.add(o);
+            }
+        }
+
+        return (isMutated) ? output : objects; // Throw away copies and use originals wherever possible
+    }
+
+    /** Sanitize a single non-Map, non-List, non-{@link Step}/non-{@link UninstantiatedDescribable} item */
+    @CheckForNull
+    Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull EnvVars vars) {
+        // Package scoped so we can test it directly
+        Object tempVal = o;
+        if (tempVal instanceof Step) {
+            // Ugly but functional used for legacy syntaxes with metasteps
```

Comment:
Huh? Under what circumstances could this happen? Even legacy metastep syntax

```groovy
step(new JUnitResultArchiver(…))
```

would be passing some `Describable`, but not a `Step`.